lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1538 @@
+    net = createNet();
+    dstNode->net = net;
+    srcNode->net = net;
+    net->nodes.insert(dstNode);
+    net->nodes.insert(srcNode);
+  } else if (!dstNode->net) {
+    // Add src into existing dst's net.
+    net = srcNode->net;
+    dstNode->net = net;
+    net->nodes.insert(dstNode);
+  } else if (!srcNode->net) {
+    // Add dst into existing src's net.
+    net = dstNode->net;
+    srcNode->net = net;
+    net->nodes.insert(srcNode);
+  } else if (srcNode->net == dstNode->net) {
+    // Both dst and src already in the same net (e.g. redundant connect).
+    net = srcNode->net;
+  } else {
+    // Use the larger of the two networks and merge the smaller one into it
+    // (wastes less space since we abandon the smaller net).
+    net = dstNode->net;
+    ResetNet *other = srcNode->net;
+    if (net->nodes.size() < other->nodes.size())
+      std::swap(net, other);
+
+    // Migrate the nodes from the other network over.
+    for (auto node : other->nodes) {
+      node->net = net;
+      net->nodes.insert(node);
+    }
+    net->drives.append(other->drives);
+    abandonNet(other);
+  }
+  assert(net);
+
+  // Add drive entry with type and loc details.
+  net->drives.push_back(ResetDrive{dstNode, srcNode, loc});
+}
+
+/// Return the reset node associated with a value. This either returns the
+/// existing node, or creates a new one if needed.
+ResetNode *ResetMap::getNode(FieldRef value, FIRRTLType type) {
+  auto &node = nodes[value];
+  if (!node)
+    node = new (allocator) ResetNode{value, type};
+  return node;
+}
+
+/// Allocate a new reset network.
+ResetNet *ResetMap::createNet() {
+  if (!unusedNets.empty()) {
+    auto *net = unusedNets.pop_back_val();
+    nets.insert(net);
+    return net;
+  }
+  auto *net = new (allocator) ResetNet();
+  allocatedNets.push_back(net);
+  nets.insert(net);
+  return net;
+}
+
+/// Abandon a reset network. This marks it as available for reuse.
+void ResetMap::abandonNet(ResetNet *net) {
+  net->clear();
+  assert(nets.remove(net));
+  unusedNets.push_back(net);
+}
+
+//===----------------------------------------------------------------------===//
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Infer concrete reset types and insert full async reset.
+///
+/// This pass replaces `reset` types in the IR with a concrete `asyncreset` or
+/// `uint<1>` depending on how the reset is used, and adds async resets to
+/// registers in modules marked with the corresponding
+/// `FullAsyncResetAnnotation`. On a high level, the pass operates as follows:
+///
+/// 1. Build a global graph of the resets in the design by tracing reset signals
+///    through instances. This uses the `ResetNetwork` utilities and is similar
+///    to establishing groups of values in the IR that are part of the same
+///    reset network (i.e., somehow attached together through ports, wires,
+///    instances, and connects).
```

Comment:
This sounds a lot like a disjoint-set data structure.

Reply:
Yeah `llvm::EquivalenceClasses` was an awesome way to simplify things!